src/cpp/include/openvino/genai/vlm_pipeline.hpp

```diff
@@ -0,0 +1,483 @@
+    imgEmbedData += embedding_dim;
+
+    if (image_embed_slices.size() > 1) {
+        //fill "<slice>" embedding
+        std::copy(data + embedding_dim * 4, data + embedding_dim * 5, imgEmbedData);
+        imgEmbedData += embedding_dim;
+
+        for (size_t i = 1; i < image_embed_slices.size(); ++i) {
+            for (size_t j = 0; j < image_embed_slices[i].size(); ++j) {
+                //fill "<image>" embedding
+                std::copy(data + embedding_dim * 2, data + embedding_dim * 3, imgEmbedData);
+                imgEmbedData += embedding_dim;
+
+                const ov::Tensor& vision_output_tensor_i_j = image_embed_slices[i][j];
+
+                //Resampler inference with OpenVINO
+                resampler.set_tensor("x", vision_output_tensor);
+                resampler.get_tensor("tgt_size").set_shape({ 1, 1 });
+                resampler.get_tensor("tgt_size").data<int64_t>()[0] = ratio;
+                resampler.get_tensor("tgt_size").data<int64_t>()[1] = ratio;
+
+                resampler.infer();
+                const ov::Tensor& vision_embded_tensor_i_j = resampler.get_output_tensor();
+                // fill image_embed_slices[i][j]
+                std::copy_n(vision_embded_tensor_i_j.data<float>(), vision_embded_tensor_i_j.get_size(), imgEmbedData);
+                imgEmbedData += n_img_pos * embedding_dim;
+
+                //fill "</image>" embedding
+                std::copy(data + embedding_dim * 3, data + embedding_dim * 4, imgEmbedData);
+                imgEmbedData += embedding_dim;
+
+                if (j == image_embed_slices[i].size() - 1) {
+                    //fill "\n" embedding
+                    std::copy(data + embedding_dim, data + embedding_dim * 1, imgEmbedData);
+                    imgEmbedData += embedding_dim;
+                }
+            }
+        }
+        //fill "</slice>" embedding
+        std::copy(data + embedding_dim * 5, data + embedding_dim * 6, imgEmbedData);
+        imgEmbedData += embedding_dim;
+    }
+    return imgEmbedding;
+}
+
+ov::Tensor process_prompt(ov::genai::Tokenizer& tokenizer, ov::InferRequest& embedding, std::string prompt) {
+    std::string user_prompt;
+    size_t idx;
+    int scale_emb = 12;
+
+    ov::Tensor input_ids = tokenizer.encode(prompt + "<AI>").input_ids;
+    auto input_len = input_ids.get_size();
+
+    ov::Tensor input_tensor = ov::Tensor(ov::element::i64, { 1, input_ids.get_size() }, input_ids.data());
+
+    embedding.set_input_tensor(input_tensor);
+    embedding.infer();
+
+    const ov::Tensor& embed_output_tensor = embedding.get_output_tensor();
+
+    ov::Shape out_shape = embed_output_tensor.get_shape();
+    float* data = embed_output_tensor.data<float>();
+
+    //embedding * scale_emb
+    for (idx = 0; idx < embed_output_tensor.get_size(); idx++) {
+        data[idx] = data[idx] * scale_emb;
+    }
+    return embed_output_tensor;
+}
+
+class VisionEncoder {
```

Comment:
will this vision encoder work for other models as well?
or it's too hardcoded for CLIP specifics